h2o-automl/src/main/java/ai/h2o/automl/targetencoding/BroadcastJoinForTargetEncoder.java

```diff
@@ -0,0 +1,137 @@
+        CompositeLookupKey s = (CompositeLookupKey) obj;
+        return _levelValue.equals(s._levelValue) && _foldValue.equals(s._foldValue);
+      }
+      return false;
+    }
+
+    @Override
+    public int hashCode() {
+      return (_levelValue + _foldValue).hashCode();
+    }
+  }
+
+  static class EncodingData {
+    private long numerator;
+    private long denominator;
+    
+    EncodingData(long numerator, long denominator) {
+      this.numerator = numerator;
+      this.denominator = denominator;
+    }
+
+    long getNumerator() {
+      return numerator;
+    }
+
+    long getDenominator() {
+      return denominator;
+    }
+  }
+
+  static class FrameWithEncodingDataToHashMap extends MRTask<FrameWithEncodingDataToHashMap> {
+
+    IcedHashMap<CompositeLookupKey, EncodingData> getEncodingDataMap() {
+      return _encodingDataMap;
+    }
+
+    IcedHashMap<CompositeLookupKey, EncodingData> _encodingDataMap = new IcedHashMap<>();
+    int _categoricalColumnIdx, _foldColumnIdx, _numeratorIdx, _denominatorIdx;
+
+    FrameWithEncodingDataToHashMap(int categoricalColumnIdx, int foldColumnId, int numeratorIdx, int denominatorIdx) {
+      this._categoricalColumnIdx = categoricalColumnIdx;
+      this._foldColumnIdx = foldColumnId;
+      this._numeratorIdx = numeratorIdx;
+      this._denominatorIdx = denominatorIdx;
+    }
+
+    @Override
+    public void map(Chunk[] cs) {
+      Chunk categoricalChunk = cs[_categoricalColumnIdx];
+      Chunk numeratorChunk = cs[_numeratorIdx];
+      Chunk denominatorChunk = cs[_denominatorIdx];
+      for (int i = 0; i < categoricalChunk.len(); i++) {
+        long levelValue = categoricalChunk.at8(i);
+        String factor = categoricalChunk.vec().factor(levelValue);
+        long foldValue = _foldColumnIdx != -1 ? cs[_foldColumnIdx].at8(i) : -1;
+        _encodingDataMap.put(new CompositeLookupKey(factor, foldValue), new EncodingData(numeratorChunk.at8(i), denominatorChunk.at8(i)));
+      }
+    }
+  }
+
+  /**
+   * 
+   * @param leftFrame frame that we want to keep order of
+   * @param leftCatColumnsIdxs indices of the categorical columns from `leftFrame` we want to use to calculate encodings
+   * @param leftFoldColumnIdx index of the fold column from `leftFrame` or `-1` if we don't use folds
+   * @param broadcastedFrame supposedly small frame that we will broadcast to all nodes and use it as lookup table for joining
+   * @param rightCatColumnsIdxs indices of the categorical columns from `broadcastedFrame` we want to use to calculate encodings 
+   * @param rightFoldColumnIdx index of the fold column from `broadcastedFrame` or `-1` if we don't use folds
+   * @return
+   */
+  static Frame join(Frame leftFrame, int[] leftCatColumnsIdxs, int leftFoldColumnIdx, Frame broadcastedFrame, int[] rightCatColumnsIdxs, int rightFoldColumnIdx) {
+    int numeratorIdx = broadcastedFrame.find("numerator");
+    int denominatorIdx = broadcastedFrame.find("denominator");
```

Comment:
not a big fan of magic strings in public classes, especially when they rely on some logic done in another public class (here TargetEncoder)

Reply:
Me neither, but in this case I would have to pass them as parameter to a main class( or worse to method) that would complicate signatures/API. But yes, these TargetEncoding and BroadcastJoiner classes are tightly coupled....but I made it explicit by adding `ForTargetEncoding` to the naming of the class.

Reply:
or just defined as constants to show where it comes from?

Reply:
@seb-h2o yep, good idea